### PR TITLE
Handle errors while receiving file changes

### PR DESCRIPTION
**Pull request by wolfs**
That is now possible with the latest native platform changes.

### CODE REVIEW

File: subprojects/core/src/main/java/org/gradle/internal/vfs/AbstractEventDrivenFileWatcherRegistry.java [hunk 1]
```diff
@@ -19,30 +19,48 @@
...
+        FileWatcher watcher = watcherCreator.createWatcher(new FileWatcherCallback() {
+            @Override
+            public void pathChanged(Type type, String path) {
+                handleEvent(type, path, handler);
+            }
+
+            @Override
+            public void reportError(Throwable ex) {
+                LOGGER.error("Error while receiving file changes", ex);
+                errorWhileReceivingFileChanges.compareAndSet(null, ex);
```
Comment (lptr) [inline subprojects/core/src/main/java/org/gradle/internal/vfs/AbstractEventDrivenFileWatcherRegistry.java]: Shall we collect all such errors in a list instead?

File: subprojects/core/src/main/java/org/gradle/internal/vfs/AbstractEventDrivenFileWatcherRegistry.java [hunk 2]
```diff
@@ -19,30 +19,48 @@
...
+                handler.handleLostState();
```
Comment (lptr) [inline subprojects/core/src/main/java/org/gradle/internal/vfs/AbstractEventDrivenFileWatcherRegistry.java]: I guess this is where we should stop the thing, no? Though calling `stop()` from here would be bad, as we are on the native thread here, and we'd get into a deadlock.

File: subprojects/core/src/main/java/org/gradle/internal/vfs/AbstractEventDrivenFileWatcherRegistry.java (hunk 1 above)
Comment (wolfs) [reply subprojects/core/src/main/java/org/gradle/internal/vfs/AbstractEventDrivenFileWatcherRegistry.java]: We could, though I am not sure it is much use. We will log the errors on the native side as well, right?

I think the first error is enough.

File: subprojects/core/src/main/java/org/gradle/internal/vfs/AbstractEventDrivenFileWatcherRegistry.java (hunk 2 above)
Comment (wolfs) [reply subprojects/core/src/main/java/org/gradle/internal/vfs/AbstractEventDrivenFileWatcherRegistry.java]: Yeah, I think stopping would make sense (or even failing the build?). Like you noted, I don't have a way to stop from right here, so what do you suggest we do?

It is also fine to keep the thing running, I think. What is the bad side effect?

Comment (lptr) [reply subprojects/core/src/main/java/org/gradle/internal/vfs/AbstractEventDrivenFileWatcherRegistry.java]: I guess it depends on what caused the problem in the first place. We might end up with logging many large exception stack traces.

File: subprojects/core/src/main/java/org/gradle/internal/vfs/AbstractEventDrivenFileWatcherRegistry.java (hunk 1 above)
Comment (lptr) [reply subprojects/core/src/main/java/org/gradle/internal/vfs/AbstractEventDrivenFileWatcherRegistry.java]: Alright.

File: subprojects/core/src/main/java/org/gradle/internal/vfs/AbstractEventDrivenFileWatcherRegistry.java (hunk 2 above)
Comment (lptr) [reply subprojects/core/src/main/java/org/gradle/internal/vfs/AbstractEventDrivenFileWatcherRegistry.java]: As to how to tear the thing down, I'm not sure. One way is to add a `terminateAsync()`-style call, returning a `Future` perhaps. We could make all the calls async, too.

**Pull request merged**